### PR TITLE
Add a consistent-crds target

### PR DESCRIPTION
**Pull request by Avni-Sharma**
Currently, crds in deploy/crds/ is not consistent with manifests/crd.yaml.
Introduction of a Makefile target `consistent-crds` - Copy crd from manifests/ and cr from examples/nodejs_postgresql to deploy/crds/ . Uses a symbolic link.
Refers https://jira.coreos.com/browse/APPSVC-188

### CODE REVIEW

**Comment by otaviof** [comment]
Do we need to execute those actions as part of `Makefile`? I think we could simply make those changes in the repository, and watch out if it changes later on.

**Comment by Avni-Sharma** [comment]
> Do we need to execute those actions as part of `Makefile`? I think we could simply make those changes in the repository, and watch out if it changes later on.

I feel this would make it easier since yaml can be cumbersome with the spaces etc.

**Comment by DhritiShikhar** [comment]
@Avni-Sharma When I run `make local` with changes in this PR, I see

```
➜  service-binding-operator git:(consistent-crd) make local
/bin/bash: line 0: cd: ./deploy/crds/: No such file or directory
make: *** [Makefile:377: consistent-crds] Error 1
```

**Comment by DhritiShikhar** [comment]
@Avni-Sharma Your PR looks good but do not delete the files under `deploy/crds`

**Comment by baijum** [comment]
@Avni-Sharma Please resolve the conflicts.

I wonder you need relative symbolic links. (Ref. https://feryn.eu/blog/relative-symlinks/)

File: Makefile [hunk 1]
```diff
@@ -394,3 +394,13 @@ else
...
 endif
+
+## -- Target for maintaining consistent crd copies --
+
+.PHONY: consistent-crds
+## Copy crd from manifests/ and cr from examples/nodejs_postgresql to deploy/crds/
+consistent-crds:
+	@cd ./deploy/crds/ && ln -srf ../../manifests-upstream/${OPERATOR_VERSION}/servicebindingrequests.apps.openshift.io.crd.yaml \
+	apps_v1alpha1_servicebindingrequest_crd.yaml
+	@cd ./deploy/crds && ln -srf ../../examples/nodejs_postgresql/service-binding-request.yaml \
```
Comment (baijum) [inline Makefile]: Use `$(Q)` variable instead of `@`

File: Makefile [hunk 2]
```diff
@@ -332,7 +332,7 @@ push-image: build-image
 
 .PHONY: local
 ## Local: Run operator locally
-local: deploy-clean deploy-rbac deploy-crds
+local: deploy-crds deploy-clean deploy-rbac
```
Comment (DhritiShikhar) [inline Makefile]: @Avni-Sharma Why did you change this order?

Comment (pratikjagrut) [reply Makefile]: @DhritiShikhar if you run `deploy-crds` target first and then `deploy-clean` it will delete the crds deployed by `deploy-crds` target. I think there isn't a problem with the original sequence. WDYT?

Comment (DhritiShikhar) [reply Makefile]: There are errors generated because of this change

```
➜  service-binding-operator git:(consistent-crds) make local
cd ./deploy/crds/ && ln -srf ../../manifests-upstream/0.0.20/servicebindingrequests.apps.openshift.io.crd.yaml \
apps_v1alpha1_servicebindingrequest_crd.yaml
cd ./deploy/crds && ln -srf ../../examples/nodejs_postgresql/service-binding-request.yaml \
apps_v1alpha1_servicebindingrequest_cr.yaml
kubectl create -f deploy/crds/apps_v1alpha1_servicebindingrequest_crd.yaml
Error from server (AlreadyExists): error when creating "deploy/crds/apps_v1alpha1_servicebindingrequest_crd.yaml": customresourcedefinitions.apiextensions.k8s.io "servicebindingrequests.apps.openshift.io" already exists
make: *** [Makefile:348: deploy-crds] Error 1
```

Comment (DhritiShikhar) [reply Makefile]: @pratikjagrut Yes. I agree with you.

@Avni-Sharma please remove this change.

File: Makefile [hunk 3]
```diff
@@ -394,3 +394,13 @@ else
...
+
+## -- Target for maintaining consistent crd copies --
+
+.PHONY: consistent-crds
+## Copy crd from manifests/ and cr from examples/nodejs_postgresql to deploy/crds/
+consistent-crds:
+	$(Q)cd ./deploy/crds/ && ln -srf ../../manifests-upstream/${OPERATOR_VERSION}/servicebindingrequests.apps.openshift.io.crd.yaml \
+	apps_v1alpha1_servicebindingrequest_crd.yaml
+	$(Q)cd ./deploy/crds && ln -srf ../../examples/nodejs_postgresql/service-binding-request.yaml \
+	apps_v1alpha1_servicebindingrequest_cr.yaml
```
Comment (pratikjagrut) [inline Makefile]: Why are crd and cr being copied from manifests and example to deploy/crds? Aren't, crds and cr from deploy/crds suppose to be original and we create manifests from it?

File: Makefile [hunk 4]
```diff
@@ -394,3 +394,13 @@ else
 		-r $(REPO_OWNER)/$(REPO_NAME) \
 		-Z > codecov-upload.log
 endif
+
+## -- Target for maintaining consistent crd copies --
+
+.PHONY: consistent-crds
+## Copy crd from manifests/ and cr from examples/nodejs_postgresql to deploy/crds/
+consistent-crds:
+	$(Q)cd ./deploy/crds/ && ln -srf ../../manifests-upstream/${OPERATOR_VERSION}/servicebindingrequests.apps.openshift.io.crd.yaml \
```
Comment (DhritiShikhar) [inline Makefile]: @Avni-Sharma We should ensure consistent crd at 3 locations:
[1] `deploy/`
[2] `manifests-upstream/`
[3] `manifests/`

There should be changes to ensure crds are consistent in `manifests/` directory as well

Comment (Avni-Sharma) [reply Makefile]: I don't think it is feasible to trigger changes to manifests/ . It should be manual because it would be taken up by the ART system.  It would automatically trigger a build for the OSBS build. wdyt @sbose78 . Should we be copying it to manifets/ as well?

Comment (Avni-Sharma) [reply Makefile]: adding another target. (to be used for OS releases)

**Comment by DhritiShikhar** [comment]
I tested the PR locally. Works as intended.

/lgtm
/approve

**Comment by sbose78** [comment]
> I don't think it is feasible to trigger changes to manifests/ . It should be manual because it would be taken up by the ART system.

+1.

> It would automatically trigger a build for the OSBS build. wdyt @sbose78 . Should we be copying it to manifests/ as well?(as soft link) (Would ART recognise soft links?)

No clue :) Yes, ideally copying over to downstream `manifests` should be less frequent.

**Comment by DhritiShikhar** [comment]
/lgtm
/approve

This PR has been hanging around for quite some time. Let's get this merged today @Avni-Sharma @sbose78
